### PR TITLE
use recover to manage panics within the builder and print a useful error

### PR DESCRIPTION
**Pull request by erikh**
/cc @tiborvass @duglin

### CODE REVIEW

**Comment by erikh** [comment]
no tests because I don't want to actually introduce a panic into the builder.

**Comment by jfrazelle** [comment]
drone no likey

**Comment by duglin** [comment]
I still feel like a newbie to "go" so I may be thinking of this wrong, but with this change I think the panic will be hidden from the daemon, right? Meaning, w/o this change I think I would see the stacktrace in the log file of the deamon. But now I don't see that, all we get is:
2015/01/08 17:43:55 Panic recieved: please report an issue with your Dockerfile attached at https://github.com/docker/docker/issues
runtime error: index out of range
on the client.  Which is a nicer error message for them, so that's good, but w/o the stacktrace on the daemon it makes it harder to debug doesn't it? Would it make sense to still print the stacktrace to the log file?

**Comment by LK4D4** [comment]
@duglin previous trace was done by code: http://golang.org/src/net/http/server.go#L1126
We can do same here I think.

File: builder/job.go [hunk 1]
```diff
@@ -27,9 +28,16 @@ func (b *BuilderJob) Install() {
 }
 
 func (b *BuilderJob) CmdBuild(job *engine.Job) engine.Status {
+	defer func() {
+		if r := recover(); r != nil {
+			job.Setenv("panic", fmt.Sprintf("Panic recieved: please report an issue with your Dockerfile attached at https://github.com/docker/docker/issues\n%s\n", r))
```
Comment (icecrime) [inline builder/job.go]: s/recieved/received

**Comment by LK4D4** [comment]
I think it is good thing to preserve printing trace.

**Comment by vbatts** [comment]
@erikh I must say, I understand why to capture panics inside the parser, but `recover()` is global and not scoped to the `CmdBuild()`. I am :-1: using `recover()` in actual runtime (or almost anywhere except for a testing sandbox).

**Comment by vbatts** [comment]
sorry, `recover()` is not global, but for the current goroutine, which would not be _so_ awful, since `CmdBuild()` is in it's own routine as a job, but still I'm not entirely keen on introducing it. If we decide to keep it, lets comment that it has been decided that its use is discouraged.

References:
* http://blog.golang.org/defer-panic-and-recover
* http://golang.org/src/encoding/json/decode.go#133

**Comment by crosbymichael** [comment]
I think we can move this into the parser, please reopen when changes are finished ( I know you are busy with many things )

**Comment by erikh** [comment]
I really would rather put it in the client. (the stack trace, that is)

**Comment by duglin** [comment]
Seems like a bit of a security risk to expose that info to the client

**Comment by erikh** [comment]
I don’t see how. If the user submitted the docker file, they know everything we can spit back at them.

-Erik

**Comment by duglin** [comment]
Showing a rogue user the server stacktrace could give them a clue as to how to perform an attack.

**Comment by icecrime** [comment]
But then the "rogue user" could clone the repo and print the stacktrace himself right?

**Comment by erikh** [comment]
I think we’re bikeshedding here. If a user has access to docker, they either have root or access to a privileged user.

-Erik

**Comment by duglin** [comment]
Not necessarily, they could be using a customized version of docker.  Either way, the stack trace should be used by the server owner, not the end user so keeping a clean separation of who sees what (when you're in a secured environment) is usually a good thing.

**Comment by duglin** [comment]
This is an interesting decision point.  I've been assuming that the docker user is not necessary the same as the docker server owner and therefore while the user could spin up containers they are not always allowed access to the daemon itself.  Which means being able to gain insight into how to, potentially, cause a DOS would want to be avoided.  But perhaps I'm alone in thinking this and everyone else is assuming the user is always trusted.

**Comment by tianon** [comment]
I would like to subscribe to your newsletter, Doug; you're making a pretty
compelling argument IMO, especially since different permission levels have
been planned as a future API goal for a very long time now.

**Comment by erikh** [comment]
Correct. They’re also vaporware atm. We can work with the now and deal with the later.

-Erik

**Comment by erikh** [comment]
Either way, I have scant time lately and would just prefer to punt this feature if there’s a significant resistance to it.

**Comment by duglin** [comment]
Just to be clear, I'm not against the recover() (I'm still too new to 'go' to know if that's the right thing to do, or the right place for it), I was just missing the stacktrace in the daemon logs.  Which I thought we could still do near wherever you put the recover() as @LK4D4 mentioned.

**Pull request closed without merging**